docs/adr/0005-support-most-recent-action-images.md

```diff
@@ -0,0 +1,29 @@
+# 5. Support of most recent action images (inc Python v2)
+
+Date: 2024-06-07
+
+## Status
+
+Accepted
+
+## Context
+
+We want to provide a Codespaces environment that's intuitive for researchers to understand and that provides a "habitable" environment for them in which to write code. We also want to keep the maintenance burden for developers low.
+
+We would also like to encourage researchers to use the newest action images available. We believe that providing a development environment that supports that aim will help with this.
+
+Note: support of action images only refers to the files we pull in from those images to provide syntax highlighting, help pages/documentation, and autocomplete. It does not affect the use of the `opensafely run` or `opensafely exec` where all versions of action images will be available as they are locally.
+
+## Decision
+
+We will provide a development environment for the most recent version of the Python Action image, which is V2.
+
+We will not provide a development environment for V1 of the Python Action image.
+
+More generally, we will aim to support the most recent versions of action images only. We will not provide support for multiple versions of an action image.
+
+## Consequences
+
+We did a soft rollout of the Python v2 action image earlier this year. At present only one or two researchers are using it, although we expect that number to increase over time. So it's possible that we're making a decision that means the Codespaces experience won't be as good for some older projects. However, we believe those numbers are very low and that the impact of not supporting Python v1 is also low. We already recommend that users run their code regularly using `opensafely run` or `opensafely exec` and those commands do still support Python v1.
+
+In future, researchers using older, unsupported action images could edit their dev container config to point to an older Docker image that still uses the older action images. However, we could potentially make other changes that means this doesn't work as expected. Over time, we may find that the Codespaces environment doesn't meet the needs of developers of older research projects. We intend to monitor the usage of Codespaces to pick up if researchers are using it on older projects. If it becomes a problem, we could change our approach.
```

Comment:
I don't think older (i.e. not the most recent) action images are unsupported. If a security vulnerability were found, then they would be updated. It's just that they aren't being updated in a backwards-incompatible way. Could you remove the word "unsupported"?

I think we should document how we'd expect a researcher to edit their dev container config to point to an older Docker image. Do we expect them to edit the following line in `devcontainer.json`?

```json
"image": "ghcr.io/opensafely-core/research-template:v0",
```

If we do, then what should it look like?

Reply:
I meant not supported by us with Codespaces rather than the action images themselves being unsupported. I've written it in a confusing way though and I agree about removing it.

Yes, I can be clearer about what editing the dev container config would look like. Good idea.